metrics/fds_linux.go

```diff
@@ -4,10 +4,109 @@
 
 package metrics
 
-import "os"
+import (
+	"bytes"
+	"fmt"
+	"log"
+	"syscall"
+	"unsafe"
+)
+
+var procSelfFDName = []byte("/proc/self/fd\x00")
 
 func currentFDs() int {
-	// TODO(bradfitz): do this without so many allocations on Linux.
-	ents, _ := os.ReadDir("/proc/self/fd")
-	return len(ents)
+	r0, _, e1 := syscall.Syscall(syscall.SYS_OPEN,
+		uintptr(unsafe.Pointer(&procSelfFDName[0])),
+		0, 0)
+	if e1 != 0 {
+		return 0
+	}
+	fd := int(r0)
+	defer syscall.Close(fd)
+
+	count := 0
+
+	const blockSize = 8 << 10
+	buf := make([]byte, blockSize) // stack-allocated; doesn't escape
+	bufp := 0                      // starting read position in buf
+	nbuf := 0                      // end valid data in buf
+	dirent := &syscall.Dirent{}
+	var err error
+	for {
+		if bufp >= nbuf {
+			bufp = 0
+			nbuf, err = readDirent(fd, buf)
+			if err != nil {
+				log.Printf("currentFDs: readDirent: %v", err)
+				return 0
```

Comment:
In the event of an error, would returning the current count be better than just reporting zero?

Reply:
I think we're more likely to notice if the graph drops to 0

Reply:
Fair argument.